hail/src/main/scala/is/hail/io/TextMatrixReader.scala

```diff
@@ -211,27 +212,29 @@ object TextMatrixReader {
 
     val opts = TextMatrixReaderOptions(params.comment, params.hasHeader)
 
-    val headerInfo = parseHeader(fs, resolvedPaths.head, separator, rowFields.size, opts)
+    val headerInfo = parseHeader(fs, fileStatuses.head.getPath, separator, rowFields.size, opts)
     if (params.addRowId && headerInfo.rowFieldNames.contains("row_id")) {
       fatal(
         s"""If no key is specified, `import_matrix_table`, uses 'row_id'
            |as the key, please provide a key or choose a different row field name.\n
            |  Row field names: ${headerInfo.rowFieldNames}""".stripMargin)
     }
     val rowFieldTypeWithoutRowId = verifyRowFields(
-      resolvedPaths.head, headerInfo.rowFieldNames, rowFields)
+      fileStatuses.head.getPath, headerInfo.rowFieldNames, rowFields)
     val rowFieldType =
       if (params.addRowId)
         TStruct("row_id" -> TInt64) ++ rowFieldTypeWithoutRowId
       else
         rowFieldTypeWithoutRowId
     if (params.hasHeader)
       warnDuplicates(headerInfo.columnIdentifiers.asInstanceOf[Array[String]])
-    val lines = HailContext.maybeGZipAsBGZip(fs, params.gzipAsBGZip) { () =>
-      val localOpts = opts
-      SparkBackend.sparkContext("TextMatrixReader.fromJValue").textFilesLines(resolvedPaths, params.nPartitions.getOrElse(backend.defaultParallelism))
-        .filter(line => line.value.nonEmpty && !localOpts.isComment(line.value))
-    }
+
+    val lines = GenericLines.read(fs, fileStatuses, params.nPartitions, None, None, params.gzipAsBGZip, false)
+      .toRDD()
+      .filter { line =>
+        val l = line.toString
```

Comment:
So `toString` is relatively expensive (constructs the string).  In my code, I pushed the filter into the the next step.  If that's not possible, options might be:
 - make GenericLine memoize toString (this is a bit subtle because it is mutable),
 - map the RDD to a (GenericLine, String) pair first
I probably prefer the first because it avoids the allocation.  Let me know if you want help with this.

Reply:
> make GenericLine memoize toString (this is a bit subtle because it is mutable),

Tried this without realizing it was mutated -- took it out.

toString is only used once here, except in `checkHeaders` which will only call it on one line per partition anyway.

Reply:
It's used again in the actual parsing of the line:

 > linesBody(lc).filter(_.toString.nonEmpty)

Reply:
yeah,  realized that. I'm memoizing the string